packages/components/src/input-control/state.js

```diff
@@ -0,0 +1,254 @@
+/**
+ * External dependencies
+ */
+import memize from 'memize';
+/**
+ * WordPress dependencies
+ */
+import { useReducer } from '@wordpress/element';
+
+const initialStateReducer = ( state ) => state;
+
+const initialInputControlState = {
+	_event: {},
+	error: null,
+	initialValue: '',
+	isDirty: false,
+	isDragEnabled: false,
+	isDragging: false,
+	isPressEnterToSubmit: false,
+	value: '',
+};
+
+const actionTypes = {
+	CHANGE: 'CHANGE',
+	DRAG_END: 'DRAG_END',
+	DRAG_START: 'DRAG_START',
+	DRAG: 'DRAG',
+	INVALIDATE: 'INVALIDATE',
+	PRESS_DOWN: 'PRESS_DOWN',
+	PRESS_ENTER: 'PRESS_ENTER',
+	PRESS_UP: 'PRESS_UP',
+	RESET: 'RESET',
+	SUBMIT: 'SUBMIT',
+	UPDATE: 'UPDATE',
+};
+
+export const inputControlActionTypes = actionTypes;
+
+/**
+ * Prepares initialState for the reducer.
+ *
+ * @param {Object} initialState The initial state.
+ * @return {Object} Prepared initialState for the reducer
+ */
+function mergeInitialState( initialState = initialInputControlState ) {
+	const { value } = initialState;
+
+	return {
+		...initialInputControlState,
+		...initialState,
+		initialValue: value,
+	};
+}
+
+/**
+ * Combines multiple stateReducers into a single stateReducer, building
+ * the pipeline to control the flow for state and actions.
+ *
+ * @param  {...any} fns State reducers.
+ * @return {Function} The single combined stateReducer.
+ */
+export const combineStateReducers = ( ...fns ) => {
+	const reducers = fns.flat( Infinity );
```

Comment:
Are we polyflling `flat`? ([MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flat#Browser_compatibility))

Reply:
I'm not sure! Is there a way for me to check? 😊 

Alternatively, we can use `flattenDeep` from lodash:
https://lodash.com/docs/4.17.15#flattenDeep

Reply:
> I'm not sure! Is there a way for me to check? 😊

Usually, a wild @aduth appears!

Reply:
It's super effective!

Reply:
🧞‍♂️ 

It appears that it is not polyfilled.

The most direct method is to test in Internet Explorer itself, most easily via combination of [VirtualBox](https://www.virtualbox.org/) and [an official VM](https://developer.microsoft.com/en-us/microsoft-edge/tools/vms/).

![image](https://user-images.githubusercontent.com/1779930/81082643-8c2fda00-8ec1-11ea-8773-88f072ad3511.png)

The most exhaustive method is to delve into the dependency tree.

`wp-polyfill` is the script we rely on to provide polyfills. It's just an alias for `@babel/polyfill`.

Currently, it's pinned to `@babel/polyfill@7.4.4` ([source](https://github.com/WordPress/wordpress-develop/blob/0206900cfff1b44af94f33f95510b18a77f5dd60/package.json#L78)). In turn, `@babel/polyfill@7.4.4` depends on `core-js@2.6.5` ([source](https://github.com/babel/babel/blob/v7.4.4/packages/babel-polyfill/package.json#L14), [relevant import source](https://unpkg.com/browse/@babel/polyfill@7.4.4/lib/index.js)). Scouring [`core-js@2.6.5`](https://github.com/zloirock/core-js/tree/v2.6.5#ecmascript-6-array) is a bit difficult, because it includes many experimental features, but not all of these are shipped with the default "package".

Most conclusively for me is that there is no `Array#flat` here:

https://github.com/zloirock/core-js/tree/v2.6.5#ecmascript-6-array

Interestingly, this version does include `Array#flatten` as an experimental Stage 1 feature ([source](https://github.com/zloirock/core-js/blob/v2.6.5/fn/array/flatten.js), [documentation](https://github.com/zloirock/core-js/tree/v2.6.5#stage-1-proposals))), but it is not bundled into the default bundle used by `@babel/polyfill`. At some point in the proposal process, it was renamed from `Array#flatten` to `Array#flat`.

Reply:
Thank you!!! In that case, I'll swap it out for the lodash flatten method for IE safety 👍

Reply:
> In that case, I'll swap it out for the lodash flatten method for IE safety 👍

It's probably the easiest option to unblock the pull request.

Separately, we could consider to bump the version of `@babel/polyfill` used in core. 

Prior art (React):

- https://github.com/WordPress/gutenberg/pull/14400
- https://core.trac.wordpress.org/ticket/48121

It might be enough to get us `Array#flat`.

At some point we're going to have to address the fact that [`@babel/polyfill` is deprecated](https://babeljs.io/docs/en/babel-polyfill). 😬

Reply:
> At some point we're going to have to address the fact that [`@babel/polyfill` is deprecated](https://babeljs.io/docs/en/babel-polyfill). 😬

🙈

Reply:
![](https://s3.amazonaws.com/pix.iemoji.com/images/emoji/apple/ios-12/256/see-no-evil-monkey.png)